field/TinyLang/Field/Evaluator.hs

```diff
@@ -144,57 +190,72 @@ evalBinOp Mul x y = return . UniConst Field $ x `mul` y
 evalBinOp Div x y = UniConst Field <$> divEval x y
 evalBinOp BAt x y = asIntegerEval x >>= fmap (UniConst Bool) . flip atEval y . fromIntegral
 
-evalStatementUni
-    :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
-    => Env (SomeUniConst f) -> Statement f -> Expr f a -> m (UniConst f a)
-evalStatementUni env (ELet (UniVar _ var) def) rest = do
-    defR <- evalExprUni env def
-    evalExprUni (insertVar var (Some defR) env) rest
-evalStatementUni env (EAssert expr) rest = do
-    exprR <- evalExpr env expr
-    if exprR
-        then evalExprUni env rest
-        else throwError $ AssertionFailedEvalError expr
+-- | Evaluate a program
+evalProgram :: (Monad m, Eq f, Field f, AsInteger f)
+    => Program f -> EvalT f m ()
+evalProgram = flip evalStatements (pure ()) . unProgram
+
+-- TODO:  Verify whether it is acutally right fold
+evalStatements :: (Monad m, Eq f, Field f, AsInteger f)
+    => Statements f -> EvalT f m () -> EvalT f m ()
+evalStatements (Statements stmts) kont = foldr evalStatement kont stmts
+
+-- | Evaluate a single statement that continue as @kont@
+evalStatement :: (Monad m, Eq f, Field f, AsInteger f)
+    => Statement f -> EvalT f m () -> EvalT f m ()
+evalStatement (ELet (UniVar _ var) def) kont = do
+    defR <- evalExprUni def
+    let someDefR = Some defR
+    modify (someDefR:)
+    withVar var someDefR kont
+evalStatement (EAssert expr) kont = do
+    exprR <- evalExpr expr
+    unless exprR $ throwError $ AssertionFailedEvalError expr
+    kont
+evalStatement (EFor uniVar@(UniVar _ var) start end body) kont
+    -- TODO:  Figure out if this is the right semantics
+    | start < end =
+        withVar var (Some $ fromIntegral start) $
+            evalStatements body $
+                evalStatement (EFor uniVar (start + 1) end body) kont
+    | otherwise =
+        let actualEnd = max start end in
+        withVar var (Some $ fromIntegral actualEnd) kont
```

Comment:
Wait, so we have python-like loops now? I.e. `for i = x to y` now means `for i in [x, y)`. That's now what they have in the Sapling spec and it's currently the single source of truth for us. Or am I missing something?

Reply:
Sorry, fixed this now.